nixos/prometheus: Support environmentFile

pkern opened the pull request:
###### Motivation for this change

For the same reason Alertmanager supports environmentFile to pass secrets along, it is useful to support the same for Prometheus' configuration to store authentication/bearer tokens outside the Nix store.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

pkern committed:
nixos/prometheus: Support environmentFile

For the same reason Alertmanager supports environmentFile to pass
secrets along, it is useful to support the same for Prometheus'
configuration to store bearer tokens outside the Nix store.

Use PrivateTmp on the systemd unit to safeguard potential secrets
in the configuration file.

mweinelt commented:
If you are going to store secrets in /tmp I think `PrivateTmp` might be beneficial.

pkern commented:
Whoops, excellent point! Fixed.

pkern commented:
@benley @fpletz @globin @WilliButz @Frostman - Would someone of you mind to take a look at this? (I can't seem to add reviewers and I don't think this will review itself by waiting.)

WilliButz reviewed (changes requested):
Thank you for you contribution :+1:
I have some suggestions for adding the environment file:
* as in the `codimd` module, I find it very important to make it clear to users, that the specified path for `environmentFile` needs to be available on the same host, the corresponding service is running on (when used with Nixops for example).
* instead of using `/tmp` with `PrivateTmp`, I'd suggest using a service-specific `RuntimeDirectory` with `RuntimeDirectoryMode` set to `0700`, an example of this is in the `murmur` module.
* to make injecting additional environment files easier for more advanced use-cases, I suggest to make `EnvironmentFile` a list of files by default (using `mkIf (cond) [ cfg.environmentFile ];`) for easier option value merging.
* also when `envsubst` was new to me and I was just given the same short description about how replacement of option values works, it didn't make it clear to me how the Nix config and environmentFile actually needed to be structured for it to work together, hence the longer description in the mentioned `codimd` module with a short example for each.

(Note that I also think the mentioned points also apply to the alertmanager module, for which I should probably open a PR soon)

pkern committed:
prometheus: implement reviewer suggestions for environmentFile support

- Move away from PrivateTmp and instead use a private RuntimeDirectory
  for the substituted config file.
- Improve documentation of environmentFile by including more descriptive
  text from existing packages (murmur and codimd).
- Also include an example scrape config entry (from nixos) as well as
  the corresponding example for the environment file.
- Pass the environment file to the systemd unit file generator as a
  list.

pkern commented:
@WilliButz Thank you for the review, that was incredibly useful. I tried to implement your changes and it seems to have worked on my NixOS test installation. Please let me know what you think.

WilliButz reviewed (approved):
Looks good! :tada:

WilliButz commented:
@GrahamcOfBorg test prometheus

WilliButz commented:
Test failed due to thanos, prometheus is working fine - they should probably be in separate tests.

WilliButz commented on nixos/modules/services/monitoring/prometheus/default.nix:
```diff
@@ -522,6 +522,45 @@ in {
+            job_name = "home_assistant";
+            metrics_path = "/api/prometheus";
+            scheme = "https";
+            bearer_token = "\${HOME_ASSISTANT_BEARER_TOKEN}";
```
OfBorg failed becaue this variable needs to be escaped with two single quotes in a multi-line string:
```suggestion
            bearer_token = "''${HOME_ASSISTANT_BEARER_TOKEN}";
```
(see the part about multi-line strings in https://nixos.org/manual/nix/stable/#idm140737322059824)

pkern replied:
I *think* I need `"\''$`. I copied it from my config and envsubst needs to see the literal $ and it's not a '' string in that config. Will test that. :)

WilliButz replied:
Oh yeah you're right, the backslash needs to be in the rendered example.

pkern committed:
prometheus: fix documentation string
